[CI] TermsAggregatorTests testOrderByCardinality failing

**Build Scans:**
- [elasticsearch-periodic #11908 / release-tests](https://gradle-enterprise.elastic.co/s/6mpg44bh7unzc)
- [elasticsearch-periodic #11908 / release-tests](https://gradle-enterprise.elastic.co/s/nl77zxxusrq6s)

**Reproduction Line:**
```
./gradlew ":server:test" --tests "org.elasticsearch.search.aggregations.bucket.terms.TermsAggregatorTests.testOrderByCardinality" -Dtests.seed=B03261B90D513D85 -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dlicense.key=x-pack/license-tools/src/test/resources/public.key -Dtests.locale=bn -Dtests.timezone=Africa/Juba -Druntime.java=25
```

**Applicable branches:**
9.3

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.search.aggregations.bucket.terms.TermsAggregatorTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testOrderByCardinality'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: <[9, 8, 7]>
     but: was <[9, 7, 8]>
```

**Issue Reasons:**
- [9.3] 2 consecutive failures in step release-tests
- [9.3] 2 failures in test testOrderByCardinality (6.9% fail rate in 29 executions)
- [9.3] 2 failures in step release-tests (66.7% fail rate in 3 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -350,9 +350,6 @@ tests:
 - class: org.elasticsearch.xpack.esql.qa.multi_node.FieldExtractorIT
   method: testTsIndexConflictingTypes {NONE}
   issue: https://github.com/elastic/elasticsearch/issues/142477
-- class: org.elasticsearch.search.aggregations.bucket.terms.TermsAggregatorTests
-  method: testOrderByCardinality
-  issue: https://github.com/elastic/elasticsearch/issues/142484
 - class: org.elasticsearch.xpack.esql.qa.single_node.GenerativeForkIT
   method: test {csv-spec:match-function.TestMatchWithReplace}
   issue: https://github.com/elastic/elasticsearch/issues/142501

--- a/server/src/main/java/org/elasticsearch/search/aggregations/metrics/HyperLogLogPlusPlus.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/metrics/HyperLogLogPlusPlus.java
@@ -490,7 +490,7 @@ public final class HyperLogLogPlusPlus extends AbstractHyperLogLogPlusPlus {
         }
 
         long maxOrd() {
-            return cells.size() - 1;
+            return cells.size();
         }
 
         @Override

--- a/server/src/test/java/org/elasticsearch/search/aggregations/metrics/HyperLogLogPlusPlusTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/metrics/HyperLogLogPlusPlusTests.java
@@ -32,6 +32,7 @@ import static org.elasticsearch.search.aggregations.metrics.AbstractCardinalityA
 import static org.hamcrest.Matchers.closeTo;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.lessThan;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
@@ -191,6 +192,24 @@ public class HyperLogLogPlusPlusTests extends ESTestCase {
             ByteSizeValue.ofBytes((initialBucketCount << precision) + initialBucketCount * 4 + PageCacheRecycler.PAGE_SIZE_IN_BYTES * 2),
             bigArrays -> new HyperLogLogPlusPlus(precision, bigArrays, initialBucketCount)
         );
+    }
+
+    public void testMaxOrdIsExclusiveUpperBound() {
+        final int p = randomIntBetween(MIN_PRECISION, MAX_PRECISION);
+        final CircuitBreaker breaker = new NoopCircuitBreaker("test");
+        // Use initialBucketCount=1 so that hll.maxOrd() stays at 1 and doesn't mask lc.maxOrd() bugs.
+        // Iterate through enough buckets to guarantee we cross at least one internal array growth boundary,
+        // where the off-by-one in LinearCounting.maxOrd() would surface.
+        try (HyperLogLogPlusPlus counts = new HyperLogLogPlusPlus(p, BigArrays.NON_RECYCLING_INSTANCE, breaker, 1)) {
+            for (int bucket = 0; bucket < 50; bucket++) {
+                counts.collect(bucket, BitMixer.mix64(bucket));
+                assertThat(
+                    "maxOrd must be an exclusive upper bound after collecting into bucket " + bucket,
+                    (long) bucket,
+                    lessThan(counts.maxOrd())
+                );
+            }
+        }
     }
 
     public void testDynamicGrowth() {

